Support using a default user(name)

mexisme opened the pull request:
FIXES #281

mexisme commented on lib/net/ssh/authentication/session.rb:
```diff
@@ -43,6 +46,7 @@ def initialize(transport, options={})
       @transport = transport
 
       @auth_methods = options[:auth_methods] || Net::SSH::Config.default_auth_methods
+      @default_user = options[:default_user] || Etc.getlogin
```
I was unsure whether the preference is reading `Etc.getlogin` directly, or somehow via a similar `Net::SSH::Config` setting?

mfazekas commented:
I'm not sure about default_user case option. What is the usecase of 

```Net::SSH.start('localhost', nil,{default_username:'foo'}) ```
vs
```Net::SSH.start('localhost', 'foo')```

??

mexisme commented:
> I'm not sure about default_user case option. What is the usecase of
> 
> Net::SSH.start('localhost', nil, {default_username:'foo'})
> vs
> Net::SSH.start('localhost', 'foo')
> 
> ??

It was a little tricky for me to follow how the default user _should_ be set:  when using the `ssh` command, it will decide the default user based on things like `~/.ssh/config` as well as the client-local username.
Should we be emulating this behaviour, or something slightly different, for example?
So, I erred on the side of flexibility for the first attempt, thinking it would be much easier to enable these other types of behvaiour if they were needed.

Also, some of the other Net::SSH options suggested that it was appropriate to allow additional overrides.

i.e. https://github.com/net-ssh/net-ssh/pull/282/files#diff-24b86dee68a1ab601927a977263f3eb2R49:
> I was unsure whether the preference is reading Etc.getlogin directly, or somehow via a similar Net::SSH::Config setting?


I overlooked explaining that, I'm sorry.   :-/

mfazekas commented:
One argument is keeping this as simple as possible. The other is that this feature is not really usefull unless we support 
```
Host example
   User foo
Host example2
   User bar
```

After all we could just ```Etc.getlogin``` in Net::SSH.start. But looking at code we already seems to support that. We just need to default to Etc.getlogin when options(:user) or passed in user is nil.
I'm thinking on changing 
https://github.com/net-ssh/net-ssh/blob/master/lib/net/ssh.rb#L225
to
```ruby
user = options.fetch(:user, user) || Etc.getlogin
```

mexisme commented:
> One argument is keeping this as simple as possible.

Sounds good.

> After all we could just Etc.getlogin in Net::SSH.start. But looking at code we already seems to support that.

Thanks, that bit was slightly unclear (and how/where it was supported).

> We just need to default to Etc.getlogin when options(:user) and the passed in user is nil.
> I'm thinking on changing 
> https://github.com/net-ssh/net-ssh/blob/master/lib/net/ssh.rb#L225 to
> 
> `user = options.fetch(:user, user) || Etc.getlogin`

I wondered that, but since the below line supports default password style, it seemed that was the better place to support default username, as well.
(at https://github.com/net-ssh/net-ssh/pull/282/files#diff-24b86dee68a1ab601927a977263f3eb2L54):
```
def authenticate(next_service, username, password=nil)
```


i.e. change (at https://github.com/net-ssh/net-ssh/pull/282/files#diff-24b86dee68a1ab601927a977263f3eb2R60):
```
  username ||= default_user
```
to:
```
username ||= Etc.getlogin
```

mfazekas commented:
>I wondered that, but since the below line supports default password style, it seemed that was the better place to support default username, as well.
>(at https://github.com/net-ssh/net-ssh/pull/282/files#diff-24b86dee68a1ab601927a977263f3eb2L54):
>
>def authenticate(next_service, username, password=nil)

I don't think  your interpretation of  ```password=nil``` as default password style is correct. For me it more reads as  "password doesn't make sense for some of the authentication methods - like hostbased and publickey".

We'll also need to revise places where the user is defauled to ```ENV["USER"]``` like:
https://github.com/net-ssh/net-ssh/blob/master/lib/net/ssh/authentication/methods/hostbased.rb#L36

mexisme commented:
> I don't think your interpretation of password=nil as default password style is correct. For me it more reads as "password doesn't make sense for some of the authentication methods - like hostbased and publickey".

Is there any liklihood that someone would want to use the Session class directly?
If so (and it looked like this might be true) then it doesn't feel right to make those users to deal with default behaviour.

> We'll also need to revise places where the user is defaulted to ENV["USER"] like:
> https://github.com/net-ssh/net-ssh/blob/master/lib/net/ssh/authentication/methods/hostbased.rb#L36

```
            def authenticate_with(identity, next_service, username, key_manager)
              debug { "trying hostbased (#{identity.fingerprint})" }
              client_username = ENV['USER'] || username
```

That's an unexpected bit of code.  Why is that being done?
On most Unix systems, that would mean the passed-in `username` would always be ignored.

mfazekas commented:
> Is there any liklihood that someone would want to use the Session class directly?
> If so (and it looked like this might be true) then it doesn't feel right to make those users to deal with default behaviour.

I don't think many people would use Net::SSH::Authentication::Session directly.  I also don't think they would be much better with 
```auth.authenticate("ssh-connection", user_might_be_nil)``` vs  calling ```auth.authenticate("ssh-connection", user_might_be_nil || Etc.getlogin)```
I'd personally favour doing the defaulting at the caller and non changing Net::SSH::Authentication::Session, just because that looks a smaller change, and there is no significant advantage to the other version.

mexisme commented:
> I don't think many people would use Net::SSH::Authentication::Session directly. I also don't think they would be much better with  auth.authenticate("ssh-connection", user_might_be_nil) vs calling auth.authenticate("ssh-connection", user_might_be_nil || Etc.getlogin)

My only thought is that this doesn't reflect default SSH behaviour, so might be surprising to some.
I may add a small comment (or doc-comment?) mentioning this, if that's OK.

>I'd personally favour doing the defaulting at the caller and non changing Net::SSH::Authentication::Session, just because that looks a smaller change, and there is no significant advantage to the other version.

OK.  I've updated the PR to reflect this -- no tests, yet, but I'm writing them.

Thanks for talking this through with me.

mexisme commented:
> > I'd personally favour doing the defaulting at the caller and non changing Net::SSH::Authentication::Session, just because that looks a smaller change, and there is no significant advantage to the other version.
> 
> OK. I've updated the PR to reflect this -- no tests, yet, but I'm writing them.

Tests written.
Doc-comments updated.

NB:  I rebased the original PR to remove the previous code, to try to avoid polluting the canonical code with irrelevant commits.

mexisme commented:
Just wondering if you're happier with this PR, now, @mfazekas ?
Thanks.

mfazekas commented:
:+1:  Thanks much, i've merged it.
